service/src/main/java/bio/terra/workspace/service/workspace/flight/WorkspaceDeleteFlight.java

```diff
@@ -73,35 +71,53 @@ public WorkspaceDeleteFlight(FlightMap inputParameters, Object applicationContex
         new DeleteGcpProjectStep(
             appContext.getCrlService(), appContext.getGcpCloudContextService()),
         cloudRetryRule);
+
+    addStep(
+        new EnsureNoWorkspaceChildrenStep(appContext.getSamService(), userRequest, workspaceUuid));
+
     addStep(
         new DeleteAzureContextStep(appContext.getAzureCloudContextService(), workspaceUuid),
         cloudRetryRule);
     addStep(
         new DeleteAwsContextStep(appContext.getAwsCloudContextService(), workspaceUuid),
         cloudRetryRule);
+    addAuthZSteps(appContext, inputParameters, userRequest, workspaceUuid, terraRetryRule);
+    addStep(
+        new DeleteWorkspaceStateStep(appContext.getWorkspaceDao(), workspaceUuid), terraRetryRule);
+  }
+
+  /**
+   * Adds steps to delete sam resources owned by WSM Workspace authz is handled differently
+   * depending on whether WSM owns the underlying Sam resource or not, as indicated by the workspace
+   * stage enum.
+   */
+  protected void addAuthZSteps(
```

Comment:
Why is this method protected?

Reply:
I was thinking of unit tests, but there's really no point: form a testing standpoint, it's all the same 'unit'.